src/main/java/org/openrewrite/hibernate/EmptyInterceptorToInterface.java

```diff
@@ -0,0 +1,91 @@
+import org.openrewrite.java.tree.TypeUtils;
+public class EmptyInterceptorToInterface extends Recipe {
+
+    private final String EMPTY_INTERCEPTOR = "org.hibernate.EmptyInterceptor";
+    private final String INTERCEPTOR = "org.hibernate.Interceptor";
+    private final String STATEMENT_INSPECTOR = "org.hibernate.resource.jdbc.spi.StatementInspector";
+    private static final AnnotationMatcher OVERRIDE_ANNOTATION_MATCHER = new AnnotationMatcher("java.lang.Override");
+
+    @Override
+    public String getDisplayName() {
+        return "Replace `extends EmptyInterceptor` with `implements Interceptor` and potentially `StatementInspector`";
+    }
+
+    @Override
+    public String getDescription() {
+        return "In Hibernate 6.0 the `Interceptor` interface received default implementations therefore the NOOP implementation that could be extended was no longer needed. " +
+               "This recipe migrates 5.x `Interceptor#onPrepareStatement(String)` to 6.0 `StatementInspector#inspect()`.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(new FindImplementations(EMPTY_INTERCEPTOR),
+                new JavaIsoVisitor<ExecutionContext>() {
+                    @Override
+                    public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext ctx) {
+                        J.ClassDeclaration cd = super.visitClassDeclaration(classDecl, ctx);
+                        if (cd.getExtends() != null && TypeUtils.isOfClassType(cd.getExtends().getType(), EMPTY_INTERCEPTOR)) {
+                            cd = cd.withExtends(null).withImplements(ListUtils.concat(cd.getImplements(), (TypeTree) TypeTree.build("Interceptor").withType(JavaType.buildType(INTERCEPTOR)).withPrefix(Space.SINGLE_SPACE)));
+                            Boolean prepareStatement = getCursor().pollMessage("prepareStatementFound");
+                            if (Boolean.TRUE.equals(prepareStatement)) {
+                                cd = cd.withImplements(ListUtils.concat(cd.getImplements(), (TypeTree) TypeTree.build("StatementInspector").withType(JavaType.buildType(STATEMENT_INSPECTOR)).withPrefix(Space.SINGLE_SPACE)));
+                            }
+                            maybeAddImport(INTERCEPTOR);
+                            maybeAddImport(STATEMENT_INSPECTOR);
+                            maybeRemoveImport(EMPTY_INTERCEPTOR);
+                        }
+                        return autoFormat(cd, ctx);
+                    }
+
+                    @Override
+                    public J.MethodDeclaration visitMethodDeclaration(J.MethodDeclaration method, ExecutionContext ctx) {
+                        J.MethodDeclaration md = super.visitMethodDeclaration(method, ctx);
+                        if (md.getLeadingAnnotations().stream().anyMatch(OVERRIDE_ANNOTATION_MATCHER::matches) && md.getSimpleName().equals("onPrepareStatement")) {
```

Comment:
Did you try using a method matcher here instead? Would be more reliable.

Reply:
Knut commented this on my other PR as well but I can't seem to get method matching against method declarations working for me, maybe we can talk about it when you get back :)